File: VB .NET/skebbyAPI.vb
Imports System
Imports System.IO
Imports System.Web
Imports System.Net
Imports System.Text
Imports System.Collections.Generic

Namespace skebbyGW

    Public Class skebbyAPI
        Protected Username As String
        Protected Password As String
        Protected Url As String

        Sub New(ByVal user As String, ByVal pass As String)
            MyClass.Username = WebUtility.UrlEncode(user)
            MyClass.Password = WebUtility.UrlEncode(pass)
            MyClass.Url = "http://gateway.skebby.it/api/send/smseasy/advanced/http.php"
        End Sub

        Protected Function do_post_request(ByVal data As Dictionary(Of String, String))
            Dim request As WebRequest = WebRequest.Create(MyClass.Url)
            request.Credentials = CredentialCache.DefaultCredentials
            CType(request, HttpWebRequest).UserAgent = "Skebby VB.NET Gateway"
            request.Method = "POST"

            Dim parameters As String = ""
            For Each item As KeyValuePair(Of String, String) In data
                parameters = parameters & "&" & item.Key & "=" & WebUtility.UrlEncode(item.Value)
            Next
            Dim byteArray As Byte() = Encoding.UTF8.GetBytes(parameters)

            request.ContentType = "application/x-www-form-urlencoded"
            request.ContentLength = byteArray.Length

            Dim dataStream As Stream = request.GetRequestStream()

            dataStream.Write(byteArray, 0, byteArray.Length)
            dataStream.Close()

            Dim response As WebResponse = request.GetResponse()

            dataStream = response.GetResponseStream()

            Dim reader As New StreamReader(dataStream)
            Dim responseFromServer As String = reader.ReadToEnd()

            reader.Close()
            dataStream.Close()
            response.Close()

            Dim result As Dictionary(Of String, String) = New Dictionary(Of String, String)

            If responseFromServer.Length <> 0 Then
                Dim values = responseFromServer.Split(New Char() {"&"c})
                For Each value In values
                    Dim temp = value.Split(New Char() {"="c})
                    result.Add(WebUtility.UrlDecode(temp(0)), WebUtility.UrlDecode(temp(1)))
                Next
            Else
                result.Add("result", "fail")
            End If

            Return result

        End Function

        Public Function sendSMS(
            ByVal text As String, _
            ByVal recipients As String(), _
            Optional ByVal method As String = "send_sms_classic", _
            Optional ByVal sender_number As String = "", _
            Optional ByVal sender_string As String = "", _
            Optional ByVal charset As String = "", _
            Optional ByVal delivery_start As String = "", _
            Optional ByVal encoding_scheme As String = "", _
            Optional ByVal validity_period As String = "", _
            Optional ByVal user_reference As String = "" _
        )
            Dim i As Integer = 0
            Dim data As New Dictionary(Of String, String)
            data.Add("username", MyClass.Username)
            data.Add("password", MyClass.Password)
            data.Add("text", text)

            For Each item As String In recipients
                data.Add("recipients[" & i & "]", item)
                i = i + 1
            Next

            data.Add("method", method)
            If (sender_number <> "") Then
                data.Add("sender_number", sender_number)
            End If
            If (sender_string <> "") Then
                data.Add("sender_string", sender_string)
            End If
            If (charset <> "") Then
                data.Add("charset", charset)
            End If
            If (delivery_start <> "") Then
                data.Add("delivery_start", delivery_start)
            End If
            If (encoding_scheme <> "") Then
                data.Add("encoding_scheme", encoding_scheme)
            End If
            If (validity_period <> "") Then
                data.Add("validity_period", validity_period)
            End If
            If (user_reference <> "") Then
                data.Add("user_reference", user_reference)
            End If

            Return do_post_request(data)
        End Function

        Public Function getCredit(
            Optional ByVal charset As String = ""
        )
            Dim data As New Dictionary(Of String, String)
            data.Add("method", "get_credit")
            data.Add("username", MyClass.Username)
            data.Add("password", MyClass.Password)
            If (charset <> "") Then
                data.Add("charset", charset)
            End If

            Return do_post_request(data)
        End Function

        Public Function addAlias(
            ByVal alias_name As String, _
            ByVal business_name As String, _
            ByVal nation As String, _
            ByVal vat_number As String, _
            ByVal taxpayer_number As String, _
            ByVal street As String, _
            ByVal city As String, _
            ByVal postcode As String, _
            ByVal contact As String, _
            Optional ByVal charset As String = ""
        )
            Dim data As New Dictionary(Of String, String)
            data.Add("method", "add_alias")
            data.Add("username", MyClass.Username)
            data.Add("password", MyClass.Password)
            data.Add("alias", alias_name)
            data.Add("business_name", business_name)
            data.Add("nation", nation)
            data.Add("vat_number", vat_number)
            data.Add("taxpayer_number", taxpayer_number)
            data.Add("street", street)
            data.Add("city", city)
            data.Add("postcode", postcode)
            data.Add("contact", contact)
            If (charset <> "") Then
                data.Add("charset", charset)
            End If

            Return do_post_request(data)
        End Function
    End Class
End Namespace

File: C# .NET/skebbyAPI.cs
// Skebby C# Gateway Class
// Skebby namespace
namespace SkebbyGW
{	
	// Import useful libraries	
	using System;
	using System.IO;
	using System.Web;
	using System.Net;
	using System.Text;
	using System.Collections.Generic;
	
	// The actual class
	public class skebbyAPI
	{	
		// declare static variable for do_post_request method
		protected static string Url {get;set;}
		protected static string Username {get;set;}
		protected static string Password {get;set;}
		
		public skebbyAPI(string User, string Pass)
		{
			Username = User;
			Password = Pass;
			Url = "http://gateway.skebby.it/api/send/smseasy/advanced/http.php";
		}
		
		// core method, sends POST request to Skebby server
		protected static Dictionary<string, string> do_post_request( Dictionary<string, string> data ){
			
			Dictionary<string, string> result = new Dictionary<string, string>();
			string[] tempResult, temp;
			
			WebRequest request = WebRequest.Create(Url);
			request.Credentials = CredentialCache.DefaultCredentials;
			((HttpWebRequest)request).UserAgent = "Skebby C# Gateway";
			request.Method = "POST";
			request.ContentType = "application/x-www-form-urlencoded";
			
			string parameters = "";
			foreach(KeyValuePair<string, string> item in data)
			{
				parameters += "&" + item.Key + "=" + WebUtility.UrlEncode( item.Value );
			}
			parameters = parameters.Substring(1);
			
			byte[] byteArray = Encoding.UTF8.GetBytes (parameters);
			request.ContentLength = byteArray.Length;
			
			Stream dataStream = request.GetRequestStream();
			dataStream.Write (byteArray, 0, byteArray.Length);
			
			WebResponse response = request.GetResponse();
			//Console.WriteLine (((HttpWebResponse)response).StatusDescription);
			dataStream = response.GetResponseStream();
			
			StreamReader reader = new StreamReader(dataStream);
			
			string responseFromServer = reader.ReadToEnd();
			
			tempResult = responseFromServer.Split('&');
			for (int i = 0; i < tempResult.Length; i++)
            {
                temp = tempResult[i].Split('=');
                result.Add(temp[0], temp[1]);
            }
			
			reader.Close();
            response.Close();	
			dataStream.Close();
			
			return result;
		}
		// sendSMS method
		public Dictionary<string, string> sendSMS(
			string[] recipients,
			string text,
			string var_method = "",
			string sender_number = "",
			string sender_string = "",
			string charset = "",
			string delivery_start = "",
			string encoding_scheme = "",
			string validity_period = "",
			string user_reference = ""
		){
			int i = 0;
			Dictionary<string, string> data = new Dictionary<string, string>();
			
			data.Add("method",(var_method != "" ) ? var_method : "send_sms_classic");
			data.Add("username",Username);
			data.Add("password",Password);
			data.Add("text",text);
			
			foreach (var number in recipients)
			{
				data.Add("recipients["+i+"]",number);
				i++;
			}
			
			if (sender_number != "") 	{ data.Add("sender_number",sender_number); }
			if (sender_string != "") 	{ data.Add("sender_string",sender_string); }
			if (charset != "") 			{ data.Add("charset",charset); }
			if (delivery_start != "") 	{ data.Add("delivery_start",delivery_start); }
			if (encoding_scheme != "")	{ data.Add("encoding_scheme",encoding_scheme); }
			if (validity_period != "")	{ data.Add("validity_period",validity_period); }
			if (user_reference != "") 	{ data.Add("user_reference",user_reference); }
			
			return do_post_request(data);
		}
		// getCredit method
		public Dictionary<string, string> getCredit(
			string charset = ""
		)
		{
			string var_method = "get_credit";
			Dictionary<string, string> data = new Dictionary<string, string>();
			
			data.Add("method",var_method);
			data.Add("username",Username);
			data.Add("password",Password);
			if (charset != "") { data.Add("charset",charset); }
			
			return do_post_request(data);
		}
		// addAlias method
		public Dictionary<string, string> addAlias(
			string alias_name,
			string business_name,
			string nation,
			string vat_number,
			string taxpayer_number,
			string street,
			string city,
			string postcode,
			string contact,
			string charset = ""
		)
		{
			string var_method = "add_alias";
			Dictionary<string, string> data = new Dictionary<string, string>();
			
			data.Add("method",var_method);
			data.Add("username",Username);
			data.Add("password",Password);
			data.Add("alias",alias_name);
			data.Add("business_name",business_name);
			data.Add("nation",nation);
			data.Add("vat_number",vat_number);
			data.Add("taxpayer_number",taxpayer_number);
			data.Add("street",street);
			data.Add("city",city);
			data.Add("postcode",postcode);
			data.Add("contact",contact);
			if (charset != "") { data.Add("charset",charset); }
			
			return do_post_request(data);
		}
		// printResult method
		public Dictionary<string, string> printResult( Dictionary<string, string> data )
		{
			foreach (KeyValuePair<string, string> item in result1)
			{
				string line = string.Format("{0}: {1}",item.Key,item.Value);
				Console.WriteLine(line);	
			}	
		}
	}
}